src/classes/CW_AudienceCSVUploadControllerTest.cls

```diff
@@ -0,0 +1,397 @@
+	}
+	
+	static testMethod void parseLFLineTerminator()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+		insert camp;
+		CW_Schedule__c sch = TestDataUtil.createSchedule(camp);
+		insert sch;
+		
+		Test.setCurrentPage(Page.CW_AudienceCSVUpload);
+		Apexpages.currentPage().getParameters().put(ConstantUtil.PAGEPARAM_CAMPAIGNID,camp.Id);
+		
+		system.assertNotEquals(ConstantUtil.LF,CSVReader.lineTerminator);
+		String stringLFCSV = groupCSVString.replace(CSVReader.lineTerminator,ConstantUtil.LF);
+		
+		Test.startTest();
+		CW_AudienceCSVUploadController uc = new CW_AudienceCSVUploadController();
+		uc.selectedSchId = sch.Id;
+		uc.csvFile = Blob.valueOf(stringLFCSV);
+		uc.filename='Somefile.csv';
+		uc.uploadCSV();
+		Test.stopTest();
+		
+		system.assertEquals(5,[select count() from CW_Audience_Group__c where Campaign__c = :camp.Id]);
+	}
+	
+	static testMethod void parseCFLineTerminator()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+		insert camp;
+		
+		CW_Schedule__c sch = TestDataUtil.createSchedule(camp);
+		insert sch;
+		
+		Test.setCurrentPage(Page.CW_AudienceCSVUpload);
+		Apexpages.currentPage().getParameters().put(ConstantUtil.PAGEPARAM_CAMPAIGNID,camp.Id);
+		
+		system.assertNotEquals(ConstantUtil.CR,CSVReader.lineTerminator);
+		String stringCRCSV = groupCSVString.replace(CSVReader.lineTerminator,ConstantUtil.CR);
+		
+		Test.startTest();
+		CW_AudienceCSVUploadController uc = new CW_AudienceCSVUploadController();
+		uc.selectedSchId = sch.Id;
+		uc.csvFile = Blob.valueOf(stringCRCSV);
+		uc.filename='Somefile.csv';
+		uc.uploadCSV();
+		Test.stopTest();
+		
+		system.assertEquals(5,[select count() from CW_Audience_Group__c where Campaign__c = :camp.Id]);
+	}
+	
+	static testMethod void parseColonSeparatedGroups()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+		insert camp;
+		CW_Schedule__c sch = TestDataUtil.createSchedule(camp);
+		insert sch;
+		
+		CW_Audience_Group__c audGr2 = TestDataUtil.createAudienceGroup(camp.Id);
+		audGr2.Name = group2;
+		audGr2.Is_Control_Group__c = false;
+		insert audGr2;
+		
+		CW_Audience_Group__c audGr5 = TestDataUtil.createAudienceGroup(camp.Id);
+		audGr5.Name = group5;
+		audGr5.Is_Control_Group__c = true;
+		insert audGr5;
+		
+		system.assertEquals(2,[select count() from CW_Audience_Group__c where Campaign__c = :camp.Id]);
+		
+		Test.setCurrentPage(Page.CW_AudienceCSVUpload);
+		Apexpages.currentPage().getParameters().put(ConstantUtil.PAGEPARAM_CAMPAIGNID,camp.Id);
+		
+		Test.startTest();
+		CW_AudienceCSVUploadController uc = new CW_AudienceCSVUploadController();
+		uc.selectedSchId = sch.Id;
+		uc.csvFile = Blob.valueOf(groupCSVString.replace(',',':'));	// replace commas with colons
+		uc.filename='Somefile.csv';
+		uc.uploadCSV();
+		Test.stopTest();
+		
+		// 3 new records were added because of the new separator
+		system.assertEquals(5,[select count() from CW_Audience_Group__c where Campaign__c = :camp.Id]);
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.